crates/sui/tests/shared_objects_version_tests.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::PathBuf;
+use sui_config::{NetworkConfig, ValidatorInfo};
+use sui_node::SuiNode;
+use sui_types::base_types::{ObjectRef, SequenceNumber};
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{CallArg, ObjectArg, TransactionEffects};
+use sui_types::object::{Object, Owner, OBJECT_START_VERSION};
+use test_utils::authority::{spawn_test_authorities, test_authority_configs};
+use test_utils::messages::move_transaction;
+use test_utils::objects::test_gas_objects;
+use test_utils::transaction::{
+    publish_package, submit_shared_object_transaction, submit_single_owner_transaction,
+};
+
+#[tokio::test]
+async fn fresh_shared_objects_get_start_version() {
+    let mut env = TestEnvironment::new().await;
+    let (_, owner) = env.create_shared_counter().await;
+    assert!(is_shared_at(&owner, OBJECT_START_VERSION));
+}
+
+#[tokio::test]
+async fn objects_transitioning_to_shared_remember_their_previous_version() {
+    let mut env = TestEnvironment::new().await;
+    let (counter, _) = env.create_counter().await;
+
+    let (counter, _) = env.increment_owned_counter(counter).await;
+    assert_ne!(counter.1, OBJECT_START_VERSION);
+
+    let (counter, owner) = env.share_counter(counter).await;
+    assert_ne!(counter.1, OBJECT_START_VERSION);
+    assert!(is_shared_at(&owner, counter.1));
+}
+
+#[tokio::test]
+async fn shared_at_version_mismatch_on_parameter() {
+    let mut env = TestEnvironment::new().await;
+    let (counter, _) = env.create_counter().await;
+
+    let (inc_counter, _) = env.increment_owned_counter(counter).await;
+    let (old_counter, old_owner) = env.share_counter(inc_counter).await;
+
+    let (new_counter, new_owner) = env
+        .increment_shared_counter(old_counter, old_counter.1)
+        .await
+        .expect("Successful shared increment");
+
+    assert_eq!(new_owner, old_owner);
+
+    // Intentionally pass the wrong object start version, the transaction should fail.
+    let fx = env
+        .increment_shared_counter(new_counter, OBJECT_START_VERSION)
```

Comment:
This isn't working at the moment -- the error it produces is:

```
Err(RpcError("Error checking transaction input objects: [NotSharedObjectError]", "Internal error"))
```

And because the validator hit an internal error, it also panics with:

```
internal error: entered unreachable code: Histogram reporting task do not stop when Histogram instance exists', crates/sui-core/src/histogram.rs:178:17
```

The `RpcError` does seem to contain the right underlying issue, but the panic is getting in the way.  Is there a more direct way to test the interaction with the `transaction_input_checker`?

Reply:
Fixing panic in https://github.com/MystenLabs/sui/pull/4883

Reply:
The fix now makes the test run cleanly (no panic)!  The error produced is still an unstructured `RpcError`, but this is something I can work with!